src/LeanCode.Contracts/NotificationTagGenerator.cs

```diff
@@ -0,0 +1,107 @@
+using System.Text;
+
+namespace LeanCode.Contracts;
+
+public static class NotificationTagGenerator
+{
+    public const string KnownTypePrefix = "!";
+
+    /// <summary>
+    /// Generates a unique tag for oncoming notifications within a specific topic based on the provided `Type`.
+    /// The generated tag is used by clients to identify notifications. There is a corresponding method in
+    /// `LeanCode.ContractsGenerator.Generation.NotificationTagGenerator` that generates a tag based on `TypeRef`.
+    /// Both methods generate the same tags.
+    /// </summary>
+    public static string Generate(Type type)
+    {
+        return type switch
+        {
+            _ when TryKnownType(type, out var name)
+                => type.GetElementType() is Type t ?
+                    $"{KnownTypePrefix}{name}{GetArgumentsString(t)}"
+                    : $"{KnownTypePrefix}{name}{GetArgumentsString(type.GetGenericArguments())}",
+            _ when type.IsGenericType => $"{type.GetSimpleName()}{GetArgumentsString(type.GetGenericArguments())}",
+            _ => type.FullName!,
+        };
+    }
+
+    private static string GetSimpleName(this Type type)
+    {
+        var typeName = type.FullName!;
+        int backtickIndex = typeName.IndexOf('`');
+
+        if (backtickIndex > 0)
+        {
+            typeName = typeName.Substring(0, backtickIndex);
+        }
+
+        return typeName;
+    }
+
+    private static string GetArgumentsString(params Type[] args)
+    {
+        var argsBuilder = new StringBuilder();
+
+        if (args.Any())
+        {
+            argsBuilder.Append('[');
+
+            foreach (var arg in args)
+            {
+                if (argsBuilder.Length > 1)
+                {
+                    argsBuilder.Append(',');
+                }
+
+                var argName = Generate(arg);
+                argsBuilder.Append(argName);
+            }
+
+            argsBuilder.Append(']');
+        }
+
+        return argsBuilder.ToString();
+    }
+
+    private static bool TryKnownType(Type type, out string? name)
+    {
+        name = type switch
+        {
+            _ when type == typeof(object) => "Object",
```

Comment:
Apart from the `object` and ones for map & array-like types, can we go with `string _ => "String"` patterns?

Reply:
I don't think so: `An expression of type 'Type' cannot be handled by a pattern of type 'string'`. Just `typeof(string)` also won't work as it's not constant value.

Reply:
Ahhh, that's right. Totally missed that this isn't value but a `Type`. Sorry!